sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,719 @@
+necessarily mean that the user expects all the code in the application to be
+runnable on that device.  Consider an application that uses some middleware
+library, where the library's header contains kernels optimized for different
+devices.  An application should be able to compile in AOT mode with this
+library without getting errors.  Therefore the AOT compiler must not fail
+simply because the middleware header contains device code for devices that
+are not being compiled for.
+
+
+### Runtime exception if device doesn't support feature
+
+When the application submits a kernel to a device via one of the kernel
+invocation commands (e.g. `parallel_for()`), the runtime must check if the
+kernel uses optional features that are not supported on that device.  If the
+kernel uses an unsupported feature, the runtime must throw a synchronous
+`errc::kernel_not_supported` exception.
+
+When doing these checks, the runtime must consider all uses of optional
+features in the kernel's static call graph, regardless of whether those uses
+are in the same translation unit as the kernel and regardless of whether those
+uses come from device code in a shared library.
+
+This exception, however, is only required for features that are exposed via a
+C++ type or function.  Examples of this include `sycl::half` or instantiating
+`sycl::atomic_ref` for a 64-bit type.  If the kernel relies on optional
+features that are more "notional" such as sub-group independent forward
+progress (`info::device::sub_group_independent_forward_progress`), no exception
+is required.
+
+To further clarify, this exception must be thrown in the following
+circumstances:
+
+* For a kernel that is not decorated with `[[sycl::requires()]]`, the exception
+  must be thrown if the kernel uses a feature that the device does not support.
+
+* For a kernel that is decorated with `[[sycl::requires()]]`, the exception
+  must be thrown if the device does not have the aspects listed in that
+  attribute.  Note that the exception must be thrown even if the kernel does
+  not actually use a feature corresponding to the aspect, and it must be
+  thrown even if the aspect does not correspond to any optional feature.
+
+* For a kernel that is decorated with `[[sycl::requires()]]`, the FE compiler
+  will mostly check (at compile time) whether the kernel uses any features that
+  are not listed in the attribute.  However, this check only results in a
+  warning, so the runtime is still responsible for throwing the exception if
+  any of the functions called by the kernel uses an optional feature that the
+  device does not support.
+
+* For a kernel that is decorated with the `[[sycl::reqd_work_group_size(W)]]`
+  or `[[sycl::reqd_sub_group_size(S)]]` attribute, the exception must be thrown
+  if the device does not support the work group size `W` or the sub-group size
+  `S`.
+
+Note that the exception must be thrown synchronously, not delayed and thrown on
+the queue's asynchronous handler.
```

Comment:
Could you please clarify why this is a requirement? 
In the current implementation it's easy to throw a synchronous exception, but it can be a
restriction for the implementation in future.

Reply:
See my response to your other comment.